runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,499 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   bool isInstanceField = node->getOpCode().isIndirect();
+   bool isAOTCompile = cg->comp()->compileRelocatableCode();
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+
+   TR::Register *scratchReg = cg->allocateRegister();
+
+   //Check if snippet has been loaded
+   if(isInstanceField && !static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->isSnippetLoaded() ||
+      (!isInstanceField && !static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->isSnippetLoaded()))
+      loadFieldWatchSnippet(cg, node, dataSnippet, dataSnippetRegister, scratchReg, isInstanceField);
+
+   TR::Register *fieldClassReg = cg->allocateRegister();
+   TR::MemoryReference *classFlagsMemRef = NULL;
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* fieldReportLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   uint8_t numOfConditions = getNumOfConditionsForReportFieldAccess(node, isResolved, isWrite,isInstanceField, cg);
+   numOfConditions++;
+   TR::RegisterDependencyConditions  *deps =  new (cg->trHeapMemory())TR::RegisterDependencyConditions(numOfConditions, numOfConditions, cg->trMemory());
+
+   TR_PPCOutOfLineCodeSection *generateReportOOL = new (cg->trHeapMemory()) TR_PPCOutOfLineCodeSection(fieldReportLabel, endLabel, cg);
+   cg->getPPCOutOfLineCodeSectionList().push_front(generateReportOOL);
+   
+   //Load fieldClass
+   if(isInstanceField)
+      {
+      generateLoadJ9Class(node, fieldClassReg, sideEffectRegister, cg);
+      }
+   else if(isResolved)
+      {
+      /* During Non-AOT compilation the fieldClass has been populated inside the dataSnippet during compilation. 
+      During AOT compilation the fieldClass must be loaded from the snippet. The fieldClass in an AOT body is invalid. */
+      if(isAOTCompile)
+         {
+         //Load Address of snippet into fieldClassReg
+         loadAddressConstant(cg, node, (uintptrj_t)dataSnippet->getSnippetLabel(),fieldClassReg);
+         generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, fieldClassReg,
+                           new (cg->trHeapMemory()) TR::MemoryReference(fieldClassReg, offsetof(J9JITWatchedStaticFieldData, fieldClass),
+                           TR::Compiler->om.sizeofReferenceAddress(), cg));
+         }
+      else
+         {
+         J9Class * fieldClass = static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->getFieldClass();
+         loadAddressConstant(cg, node, (uintptrj_t)fieldClass, fieldClassReg);
```

Comment:
Similar to the above line, can we use C++ style casts here for `(uintptrj_t)`?

Similarly everywhere else.